File: lib/rag/chunker.ts
export type Chunk = {
  content: string;
  index: number;
  tokenCount: number;
};

export type ChunkOptions = {
  /** Target max tokens per chunk (default: 512) */
  maxTokens: number;
  /** Overlap ratio between consecutive chunks, 0-1 (default: 0.15) */
  overlap: number;
  /** Document title to prepend as context */
  documentTitle?: string;
  /** Section header to prepend as context */
  sectionHeader?: string;
};

/**
 * Approximate token count using char/4 heuristic.
 * Good enough for text-embedding-3-small; swap for tiktoken if precision needed.
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

/**
 * Split text at paragraph boundaries (double newline).
 */
function splitParagraphs(text: string): string[] {
  return text
    .split(/\n\n+/)
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
}

/**
 * Split text at sentence boundaries.
 */
function splitSentences(text: string): string[] {
  // Split on sentence-ending punctuation followed by space or end
  const sentences = text.match(/[^.!?]+[.!?]+(?:\s|$)|[^.!?]+$/g);
  if (!sentences) return [text];
  return sentences.map((s) => s.trim()).filter((s) => s.length > 0);
}

/**
 * Recursively split a segment that exceeds maxTokens.
 * Priority: paragraphs → sentences → hard split at word boundary.
 */
function splitSegment(text: string, maxTokens: number): string[] {
  if (estimateTokens(text) <= maxTokens) {
    return [text];
  }

  // Try paragraph split first
  const paragraphs = splitParagraphs(text);
  if (paragraphs.length > 1) {
    return mergeSegments(paragraphs, maxTokens);
  }

  // Try sentence split
  const sentences = splitSentences(text);
  if (sentences.length > 1) {
    return mergeSegments(sentences, maxTokens);
  }

  // Hard split at word boundary as last resort
  const words = text.split(/\s+/);
  return mergeSegments(words, maxTokens);
}

/**
 * Greedily merge small segments into chunks up to maxTokens.
 * Recursively split any individual segment that's still too large.
 */
function mergeSegments(segments: string[], maxTokens: number): string[] {
  const result: string[] = [];
  let current: string[] = [];
  let currentTokens = 0;

  for (const segment of segments) {
    const segTokens = estimateTokens(segment);

    // If a single segment exceeds maxTokens, recursively split it
    if (segTokens > maxTokens) {
      // Flush current buffer
      if (current.length > 0) {
        result.push(current.join(" "));
        current = [];
        currentTokens = 0;
      }
      // Recursively split the large segment
      const subChunks = splitSegment(segment, maxTokens);
      result.push(...subChunks);
      continue;
    }

    // Adding this segment would exceed limit — flush and start new chunk
    if (currentTokens + segTokens > maxTokens && current.length > 0) {
      result.push(current.join(" "));
      current = [];
      currentTokens = 0;
    }

    current.push(segment);
    currentTokens += segTokens;
  }

  if (current.length > 0) {
    result.push(current.join(" "));
  }

  return result;
}

/**
 * Apply overlap: take the last `overlapTokens` worth of text from the
 * previous chunk and prepend it to the current chunk.
 */
function applyOverlap(
  chunks: string[],
  overlapRatio: number,
  maxTokens: number
): string[] {
  if (chunks.length <= 1 || overlapRatio <= 0) return chunks;

  const overlapTokens = Math.floor(maxTokens * overlapRatio);
  const result: string[] = [chunks[0]];

  for (let i = 1; i < chunks.length; i++) {
    const prevWords = chunks[i - 1].split(/\s+/);
    // Take last N words that approximate overlapTokens
    let overlapText = "";
    const words: string[] = [];
    for (let j = prevWords.length - 1; j >= 0; j--) {
      const candidate = [prevWords[j], ...words].join(" ");
      if (estimateTokens(candidate) > overlapTokens) break;
      words.unshift(prevWords[j]);
      overlapText = candidate;
    }

    if (overlapText.length > 0) {
      result.push(overlapText + " " + chunks[i]);
    } else {
      result.push(chunks[i]);
    }
  }

  return result;
}

/**
 * Build the context prefix from document title and section header.
 */
function buildPrefix(
  documentTitle?: string,
  sectionHeader?: string
): string {
  const parts: string[] = [];
  if (documentTitle) parts.push(documentTitle);
  if (sectionHeader) parts.push(sectionHeader);
  if (parts.length === 0) return "";
  return parts.join(" > ") + "\n\n";
}

/**
 * Chunk text using recursive splitting strategy.
 *
 * Splits at paragraph boundaries first, then sentence boundaries,
 * then word boundaries. Applies overlap between consecutive chunks.
 * Optionally prepends document title and section header as context.
 */
export function chunkText(text: string, options: ChunkOptions): Chunk[] {
  const trimmed = text.trim();
  if (trimmed.length === 0) return [];

  const prefix = buildPrefix(options.documentTitle, options.sectionHeader);
  const prefixTokens = estimateTokens(prefix);
  const contentMaxTokens = options.maxTokens - prefixTokens;

  // Reserve space for overlap so chunks stay within maxTokens after overlap is applied
  const overlapBudget = Math.floor(contentMaxTokens * options.overlap);
  const splitTarget = contentMaxTokens - overlapBudget;

  // Split into raw chunks (smaller to leave room for overlap)
  let rawChunks = splitSegment(trimmed, splitTarget);

  // Apply overlap (prepend tail of previous chunk to each subsequent chunk)
  rawChunks = applyOverlap(rawChunks, options.overlap, contentMaxTokens);

  // Build final chunks with prefix and metadata
  return rawChunks.map((content, index) => {
    const fullContent = prefix + content;
    return {
      content: fullContent,
      index,
      tokenCount: estimateTokens(fullContent),
    };
  });
}

File: services/ingestion/src/chunker.py
import math
import re
from dataclasses import dataclass


@dataclass
class Chunk:
    content: str
    index: int
    token_count: int


@dataclass
class ChunkOptions:
    max_tokens: int
    overlap: float
    document_title: str | None = None
    section_header: str | None = None


def estimate_tokens(text: str) -> int:
    """Approximate token count: ~4 chars per token."""
    return math.ceil(len(text) / 4)


def _split_paragraphs(text: str) -> list[str]:
    return [p.strip() for p in re.split(r"\n\n+", text) if p.strip()]


def _split_sentences(text: str) -> list[str]:
    sentences = re.findall(r"[^.!?]+[.!?]+(?:\s|$)|[^.!?]+$", text)
    if not sentences:
        return [text]
    return [s.strip() for s in sentences if s.strip()]


def _merge_segments(segments: list[str], max_tokens: int) -> list[str]:
    result: list[str] = []
    current: list[str] = []
    current_tokens = 0

    for segment in segments:
        seg_tokens = estimate_tokens(segment)
        if seg_tokens > max_tokens:
            if current:
                result.append(" ".join(current))
                current = []
                current_tokens = 0
            result.extend(_split_segment(segment, max_tokens))
            continue
        if current_tokens + seg_tokens > max_tokens and current:
            result.append(" ".join(current))
            current = []
            current_tokens = 0
        current.append(segment)
        current_tokens += seg_tokens

    if current:
        result.append(" ".join(current))
    return result


def _split_segment(text: str, max_tokens: int) -> list[str]:
    if estimate_tokens(text) <= max_tokens:
        return [text]

    paragraphs = _split_paragraphs(text)
    if len(paragraphs) > 1:
        return _merge_segments(paragraphs, max_tokens)

    sentences = _split_sentences(text)
    if len(sentences) > 1:
        return _merge_segments(sentences, max_tokens)

    words = text.split()
    if len(words) > 1:
        return _merge_segments(words, max_tokens)

    # Base case: single token too large to split further — hard character split
    char_limit = max_tokens * 4  # ~4 chars per token
    return [text[i : i + char_limit] for i in range(0, len(text), char_limit)]


def _apply_overlap(chunks: list[str], overlap_ratio: float, max_tokens: int) -> list[str]:
    if len(chunks) <= 1 or overlap_ratio <= 0:
        return chunks

    overlap_tokens = int(max_tokens * overlap_ratio)
    result = [chunks[0]]

    for i in range(1, len(chunks)):
        prev_words = chunks[i - 1].split()
        overlap_words: list[str] = []

        for j in range(len(prev_words) - 1, -1, -1):
            candidate = [prev_words[j]] + overlap_words
            if estimate_tokens(" ".join(candidate)) > overlap_tokens:
                break
            overlap_words = candidate

        overlap_text = " ".join(overlap_words)
        if overlap_text:
            result.append(overlap_text + " " + chunks[i])
        else:
            result.append(chunks[i])

    return result


def _build_prefix(document_title: str | None, section_header: str | None) -> str:
    parts = [p for p in [document_title, section_header] if p]
    if not parts:
        return ""
    return " > ".join(parts) + "\n\n"


def chunk_text(text: str, options: ChunkOptions) -> list[Chunk]:
    trimmed = text.strip()
    if not trimmed:
        return []

    prefix = _build_prefix(options.document_title, options.section_header)
    prefix_tokens = estimate_tokens(prefix)
    content_max_tokens = options.max_tokens - prefix_tokens
    overlap_budget = int(content_max_tokens * options.overlap)
    split_target = content_max_tokens - overlap_budget

    raw_chunks = _split_segment(trimmed, split_target)
    raw_chunks = _apply_overlap(raw_chunks, options.overlap, content_max_tokens)

    return [
        Chunk(
            content=prefix + content,
            index=i,
            token_count=estimate_tokens(prefix + content),
        )
        for i, content in enumerate(raw_chunks)
    ]
